### DISCUSSION TITLE
Question about rust crate biome_js_semantic

### DISCUSSION THREAD

**Discussion by fireairforce** [biomejs/biome#2059] [Q&A] [answered]
The source code is: https://github.com/biomejs/biome/blob/main/crates/biome_js_semantic/src/events.rs#L555

I would like to ask why the semantics of biome will filter out the kind TS_QUALIFIED_NAME when analyzing the reference of binding. I have such a scenario of mutation.rename_node_declaration:

```ts
namespace A {
  export type B = Number;
  export const a = 1;
}

let a: A.B = A.a;
```

I want to renaming the namespace binding `A`  to `A_1` , for the reason references filter out TS_QUALIFIED_NAME, the renaming output will be:

```ts
namespace A_1 {
  export type B = Number;
  export const a = 1;
}

let a: A.B = A_1.a;
```

the `A.B` will not change here.

I'm not sure why biome does this. Is there any background reason I can understand? 

looking forward to your reply~

**Comment by Conaclos** [comment]
> I would like to ask why the semantics of biome will filter out the kind TS_QUALIFIED_NAME when analyzing the reference of binding.

Because `TsQualifiedName` may recursively include multiple qualification such as `A.B.C.D`.

The code skips `TS_QUALIFIED_NAME` and should end with a `TS_REFERENCE_TYPE` node.

Here `A.B` in `let a: A.B` is a `TsReferenceType` that includes a `TsQualifiedName`. `A` is a `JsReferenceIdentifier`. You can use the [Playground](https://biomejs.dev/playground/?code=bgBhAG0AZQBzAHAAYQBjAGUAIABBACAAewAKACAAIABlAHgAcABvAHIAdAAgAHQAeQBwAGUAIABCACAAPQAgAE4AdQBtAGIAZQByADsACgAgACAAZQB4AHAAbwByAHQAIABjAG8AbgBzAHQAIABhACAAPQAgADEAOwAKAH0ACgAKAGwAZQB0ACAAYQA6ACAAQQAuAEIAIAA9ACAAQQAuAGEAOwA%3D) to visualize the AST.

I think the semantic model correctly binds this reference to the namespace declaration.
The issue is certainly coming from the rename facility.

**Reply by fireairforce** [reply]
But the source code of `rename_node_declaration`: https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/src/utils/rename.rs#L273 : 

it use `prev_binding.all_references(model).collect()` to get `binding's` reference, and i found result of `all_references` : 

the source code: https://github.com/biomejs/biome/blob/main/crates/biome_js_semantic/src/semantic_model/binding.rs#L83 

It was created by `biome_js_semantic`'s `builder.rs`.

Maybe I can submit a pr to fix this situation with `rename_module_declaration`?

**Reply by Conaclos** [reply]
You are welcomed!

**Reply by fireairforce** [reply]
I recently debugged the `biome_js_semantic` package and found some problems:

It looks like that `biome_js_semantic` will deal the case: `A.B` which i mentioned above into `BindingName::Type`: 

https://github.com/biomejs/biome/blob/main/crates/biome_js_semantic/src/events.rs#L558-L560

```rust
self.push_reference(BindingName::Type(name), Reference::Read(range));
```

Then this `Type` will be processed into `UnresolvedReference`, which means that `A.B` will not be `namespace A`'s reference:

https://github.com/biomejs/biome/blob/main/crates/biome_js_semantic/src/events.rs#L800-L803

So the `rename_module_declaration` will not get the references for `namespace A`, so the `A.B` will not change here.

@Conaclos I’m not sure if this is due to the design of Biome itself.

**Reply by fireairforce** [reply]
If I want to fix here, may be i can deal the `A.B`(A) into  `BindingName::Value` ?

**Reply by Conaclos** [reply]
Indeed, you are right.

Unfortunately I don't see an easy fix there.
Because [this code](https://github.com/biomejs/biome/blob/main/crates/biome_js_semantic/src/events.rs#L553-L555) is useful in at least three cases:

- to skip import type qualifier. For example:
  ```ts
  type T = import("mod").A.B.C;
  ```
  Here `A.B.C` is a `TS_QULIFIED_NAME`.
- to correctly handle typeof. For example:
  ```ts
  import type { A } from "mod"
  type T = typeof A.B.C;
  ```
- to handle references to a namespace imported as a type:
  ```ts
  import type { A } from "./ns.js"
  type T = A.B;
  ```

Fixing the issue could require to use the sane trick that `BindingName::Typeof` by delaying the resolution to `pop_scope`.

**Reply by Conaclos** [reply]
Another possibility could be to add a special case in `pop_scope`: When a `TsRefrenceType` is unresolved and is a qualified name and a namespace declaration is in the scope, then bind the ref to the declaration.

**Reply by fireairforce** [reply, marked as answer]
I add a draft pr for this change: https://github.com/biomejs/biome/pull/2092, maybe you can take a look when you feel free, thank you~
